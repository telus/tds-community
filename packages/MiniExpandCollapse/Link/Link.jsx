import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { buttons } from '@tds/shared-styles'
import { small, medium, large } from '@tds/shared-typography'
import { colorGreyShark, colorWhite, colorGainsboro, colorGreyRaven } from '@tds/core-colours'
import { CaretDown, CaretUp } from '@tds/core-interactive-icon'
import { responsiveProps } from '@tds/util-prop-types'
import A11yContent from '@tds/core-a11y-content'
import { handleResponsiveStyles } from '@tds/util-helpers'
import createIconSizeProp from '../createIconSizeProp'

const states = ({ invert }) => {
  return {
    '&:hover': {
      textDecoration: 'none',
    },
    '&:active': {
      color: invert && colorGainsboro,
      backgroundColor: invert ? 'rgba(0,0,0,0.4)' : colorGainsboro,
      borderRadius: '0.25rem',
      padding: '0.125rem',
      margin: '-0.125rem',
      textDecoration: 'underline',
    },
    '&:focus': {
      border: `0.125rem solid ${invert ? colorWhite : colorGreyRaven}`,
      padding: '0.125rem',
      margin: '-0.25rem',
      borderRadius: '0.25rem',
      outline: 'none',
    },
  }
}

const titleSize = props =>
  handleResponsiveStyles({ size: props.size }, ({ size }) => {
    if (size === 'small') {
      return small
    } else if (size === 'medium') {
      return medium
    }
    return large
  })

const LinkButton = styled.button(
  buttons.noStyle,
  states,
  titleSize,
  {
    color: colorGreyShark,
    textDecoration: 'underline',
    textAlign: 'left',
    '& > svg': {
      verticalAlign: 'middle',
      marginLeft: '0.25rem',
      marginRight: '-0.25rem',
    },
  },
  ({ invert, context }) => {
    if (context.inheritColor) {
      return {
        color: 'inherit',
      }
    }
    if (invert) {
      return {
        color: colorWhite,
      }
    }
    return {}
  }
)

const Link = (
  { expandTitle, collapseTitle, isPanelOpen, togglePanel, size, invert, a11yLabel },
  context
) => {
  const handleClick = e => {
    e.preventDefault()
    togglePanel()
  }

  return (
    <LinkButton
      type="button"
      onClick={handleClick}
      size={size}
      invert={invert}
      context={context}
      a11yLabel={a11yLabel}
    >
      {isPanelOpen && expandTitle ? expandTitle : collapseTitle}
      {a11yLabel && <A11yContent>{a11yLabel}</A11yContent>}
      {isPanelOpen ? (
        <CaretUp variant={invert ? 'inverted' : 'basic'} size={createIconSizeProp(size)} />
      ) : (
        <CaretDown variant={invert ? 'inverted' : 'basic'} size={createIconSizeProp(size)} />
      )}
    </LinkButton>
  )
}

Link.propTypes = {
  expandTitle: PropTypes.string,
  collapseTitle: PropTypes.string.isRequired,
  isPanelOpen: PropTypes.bool.isRequired,
  togglePanel: PropTypes.func.isRequired,
  size: responsiveProps(PropTypes.oneOf(['small', 'medium', 'large'])).isRequired,
  invert: PropTypes.bool,
  a11yLabel: PropTypes.string,
}

Link.defaultProps = {
  expandTitle: null,
  invert: undefined,
  a11yLabel: undefined,
}

export default Link
