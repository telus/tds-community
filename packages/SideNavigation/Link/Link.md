### Minimal usage

A `SideNavigation.Link` is a navigational element that is used on its own and/or within a `SideNavigation.SubMenu`.

### Usage criteria

- `SideNavigation.Link` changes text and background to purple on hover
- `SideNavigation.Link` will put a left border on active states and will bold the text.
