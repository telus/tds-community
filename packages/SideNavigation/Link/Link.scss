@import '~@tds/core-colours/colours';

.link {
  color: black;
  text-decoration: none;
  background: none !important;
}

.active {
  border-left: 4px solid $color-telus-purple;
  color: $color-telus-purple;
}

.activeText {
  font-weight: bold;
  color: $color-telus-purple !important;
}

.box {
  &:hover {
    background-color: $color-white-lilac;
  }
}

.hover {
  color: $color-shuttle-grey;
  &:hover {
    color: $color-telus-purple;
  }
}
