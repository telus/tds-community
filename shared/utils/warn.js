const warn = (componentName, message) => {
  if (process.env.NODE_ENV === 'production') {
    return
  }

  console.warn(`[TDS] ${componentName}: ${message}`) // eslint-disable-line no-console
}

export default warn
