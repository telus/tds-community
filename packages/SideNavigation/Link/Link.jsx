import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Box from '@tds/core-box'
import Text from '@tds/core-text'
import safeRest from '@tds/shared-safe-rest'
import { colorWhiteLilac, colorTelusPurple, colorShuttleGrey } from '@tds/core-colours'

import ColoredTextProvider from '../../../shared/components/ColoredTextProvider/ColoredTextProvider'
import warn from '../../../shared/utils/warn'

const activeContainer = {
  borderLeft: `4px solid ${colorTelusPurple}`,
  color: `${colorTelusPurple}`,
}

const activeText = {
  fontWeight: 'bold',
  color: `${colorTelusPurple}`,
}

const hover = {
  color: `${colorShuttleGrey}`,
  '&:hover': {
    color: `${colorTelusPurple}`,
  },
}

const BoxContainer = styled(Box)(props => ({
  '&:hover': {
    backgroundColor: `${colorWhiteLilac}`,
  },
  ...(props.active ? activeContainer : undefined),
}))

const StyledTextProvider = styled(ColoredTextProvider)(props => ({
  ...(props.active ? activeText : hover),
}))

const StyledAnchor = styled.a({
  textDecoration: 'none',
})

/**
 * _This component can only be accessed as a name-spaced component: `SideNavigation.Link`._
 */
const Link = ({ reactRouterLinkComponent, children, active, subMenuLink, ...rest }) => {
  if ((reactRouterLinkComponent || rest.to) && !(reactRouterLinkComponent && rest.to)) {
    warn('Link', 'The props `reactRouterLinkComponent` and `to` must be used together.')
  }

  const innerLink = (
    <BoxContainer vertical={subMenuLink ? 2 : 3} horizontal={3} active={active ? 1 : 0}>
      <StyledTextProvider active={active}>
        <Text size={subMenuLink ? 'small' : 'medium'} bold={active}>
          {children}
        </Text>
      </StyledTextProvider>
    </BoxContainer>
  )

  return React.createElement(
    reactRouterLinkComponent || StyledAnchor,
    {
      ...safeRest(rest),
    },
    innerLink
  )
}

Link.propTypes = {
  /**
   * Text of the Link.
   */
  children: PropTypes.string.isRequired,
  /**
   * Target URL.
   */
  href: PropTypes.string,
  /**
   * Active state for the link (if you are on the current URL of the link).
   */
  active: PropTypes.bool,
  /**
   * Describes whether the link is used as a SubMenu header or not.
   *
   * @ignore
   */
  subMenuLink: PropTypes.bool,
  /**
   * 	React Router Link component.
   */
  reactRouterLinkComponent: PropTypes.func,
  /**
   * Target URL (if using 'reactRouterLinkComponent').
   */
  to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
}

Link.defaultProps = {
  href: undefined,
  active: false,
  subMenuLink: false,
  reactRouterLinkComponent: null,
  to: null,
}

export default Link
