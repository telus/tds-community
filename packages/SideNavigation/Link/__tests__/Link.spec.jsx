import React from 'react'
import { shallow } from 'enzyme'

import warn from '../../../../shared/utils/warn'
import Link from '../Link'

jest.mock('../../../../shared/utils/warn')

describe('SideNavigation.Link', () => {
  const children = <Link href="#">Home</Link>
  const doShallow = () => shallow(children)

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders', () => {
    const link = doShallow()

    expect(link).toMatchSnapshot()
  })

  it('renders with active styles', () => {
    const link = shallow(
      <Link href="#" active>
        Home
      </Link>
    )

    expect(link).toMatchSnapshot()
  })

  it('must use `reactRouterLinkComponent` and `to` props together', () => {
    const doShallowLink = (overrides = {}) => shallow(<Link {...overrides}>Go home</Link>)
    const MyLink = () => <span />
    doShallowLink({ reactRouterLinkComponent: MyLink })

    expect(warn).toHaveBeenCalled()

    jest.clearAllMocks()

    const link = doShallowLink({ to: '/about' })

    expect(link).toHaveProp('to')
    expect(warn).toHaveBeenCalled()
  })
})
