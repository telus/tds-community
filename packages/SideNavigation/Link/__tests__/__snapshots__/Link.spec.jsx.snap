// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavigation.Link renders 1`] = `
.c1:hover {
  background-color: #f2eff4;
}

.c2 {
  color: #54595f;
}

.c2:hover {
  color: #4b286d;
}

.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Link
  active={false}
  href="#"
  reactRouterLinkComponent={null}
  subMenuLink={false}
  to={null}
>
  <styled.a
    href="#"
    to={null}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "text-decoration: none;",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      href="#"
      to={null}
    >
      <a
        className="c0"
        href="#"
        to={null}
      >
        <Styled(Box)
          active={0}
          horizontal={3}
          vertical={3}
        >
          <StyledComponent
            active={0}
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    [Function],
                  ],
                },
                "displayName": "Styled(Box)",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": [Function],
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            horizontal={3}
            vertical={3}
          >
            <Box
              active={0}
              className="c1"
              horizontal={3}
              inline={false}
              tag="div"
              vertical={3}
            >
              <div
                active={0}
                className="TDS_Box-modules__horizontalPadding-3___2uoUp TDS_Box-modules__verticalPadding-3___Fsv37 c1"
              >
                <Styled(ColoredTextProvider)
                  active={false}
                >
                  <StyledComponent
                    active={false}
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bwzfXH",
                          "isStatic": false,
                          "lastClassName": "c2",
                          "rules": Array [
                            [Function],
                          ],
                        },
                        "displayName": "Styled(ColoredTextProvider)",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bwzfXH",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <ColoredTextProvider
                      active={false}
                      className="c2"
                    >
                      <div
                        className="c2"
                      >
                        <Text
                          block={false}
                          bold={false}
                          invert={false}
                          size="medium"
                        >
                          <span
                            className="TDS_Typography-modules__default___KlcpP TDS_Typography-modules__medium___1rxfE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__mediumFont___XMrRj TDS_Typography-modules__inheritColor___SREX0"
                          >
                            Home
                          </span>
                        </Text>
                      </div>
                    </ColoredTextProvider>
                  </StyledComponent>
                </Styled(ColoredTextProvider)>
              </div>
            </Box>
          </StyledComponent>
        </Styled(Box)>
      </a>
    </StyledComponent>
  </styled.a>
</Link>
`;

exports[`SideNavigation.Link renders with active styles 1`] = `
.c1 {
  border-left: 4px solid #4b286d;
  color: #4b286d;
}

.c1:hover {
  background-color: #f2eff4;
}

.c2 {
  font-weight: bold;
  color: #4b286d;
}

.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Link
  active={true}
  href="#"
  reactRouterLinkComponent={null}
  subMenuLink={false}
  to={null}
>
  <styled.a
    href="#"
    to={null}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "text-decoration: none;",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      href="#"
      to={null}
    >
      <a
        className="c0"
        href="#"
        to={null}
      >
        <Styled(Box)
          active={1}
          horizontal={3}
          vertical={3}
        >
          <StyledComponent
            active={1}
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    [Function],
                  ],
                },
                "displayName": "Styled(Box)",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": [Function],
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            horizontal={3}
            vertical={3}
          >
            <Box
              active={1}
              className="c1"
              horizontal={3}
              inline={false}
              tag="div"
              vertical={3}
            >
              <div
                active={1}
                className="TDS_Box-modules__horizontalPadding-3___2uoUp TDS_Box-modules__verticalPadding-3___Fsv37 c1"
              >
                <Styled(ColoredTextProvider)
                  active={true}
                >
                  <StyledComponent
                    active={true}
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bwzfXH",
                          "isStatic": false,
                          "lastClassName": "c2",
                          "rules": Array [
                            [Function],
                          ],
                        },
                        "displayName": "Styled(ColoredTextProvider)",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bwzfXH",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <ColoredTextProvider
                      active={true}
                      className="c2"
                    >
                      <div
                        className="c2"
                      >
                        <Text
                          block={false}
                          bold={true}
                          invert={false}
                          size="medium"
                        >
                          <span
                            className="TDS_Typography-modules__default___KlcpP TDS_Typography-modules__medium___1rxfE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__boldFont___2YS3O TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__inheritColor___SREX0"
                          >
                            Home
                          </span>
                        </Text>
                      </div>
                    </ColoredTextProvider>
                  </StyledComponent>
                </Styled(ColoredTextProvider)>
              </div>
            </Box>
          </StyledComponent>
        </Styled(Box)>
      </a>
    </StyledComponent>
  </styled.a>
</Link>
`;
